pht_federated/aggregator/services/discovery/column_harmonization.py

```diff
@@ -0,0 +1,217 @@
+            "hint": f"Column name {diff[0]} only exists in local dataset",
+        }
+        mismatch_errors_list.append(case)
+
+    for diff in column_value_differences2:
+        case = {
+            "column_name": diff[0],
+            "error": {
+                "type": "missing",  # missing, type, semantic, extra
+                "aggregator_type": diff[1],
+            },
+            "hint": f"Column name {diff[0]} only exists in aggregator dataset",
+        }
+        mismatch_errors_list.append(case)
+
+    for diff in column_name_differences:
+        case = {
+            "column_name": diff[0][0],
+            "error": {
+                "type": "added",  # missing, type, semantic, extra
+                "dataset_name": diff[0][0],
+                "aggregator_name": diff[1][0],
+                "aggregator_type": diff[1][1],
+            },
+            "hint": f"Column name {diff[0][0]} only exists in local dataset."
+            f" Did you mean column name: {diff[1][0]}",
+        }
+        mismatch_errors_list.append(case)
+
+    if len(mismatch_errors_list) == 0:
+        difference_report["status"] = "passed"
+    else:
+        difference_report["status"] = "failed"
+
+    difference_report["errors"] = mismatch_errors_list
+
+    return difference_report
+
+
+def intersection_two_lists(
+    df_col_names: List[Tuple[str, str]],
+    aggregator_col_names: List[Tuple[str, str]],
+):
+    """
+    Compares the column_names and types of the local dataset and the aggregated dataset and returns the intersection
+    between both
+    :param df_col_names: List[Tuple[col_name, type]]
+    :param aggregator_col_names: List[Tuple[col_name, type]]
+    :return: intersection -> List[Tuple[col_name, type]] lists column_names & types of identical columns between
+                             local dataset and aggregator dataset
+    :return: type_differences -> List[List[Tuple[col_name_df, type], Tuple[col_name_aggregator, type]]] lists
+                                 differences between local dataset and aggregator dataset if only type mismatches
+    :return: aggregator_col_names -> List[Tuple[col_name, type]] updated list that removed columns with type differences
+    :return: df_col_names -> List[Tuple[col_name, type]] updated list that removed columns with type differences
+    """
+
+    intersection = []
+    type_differences = []
+
+    for stats_keys in df_col_names:
+        for aggregator_keys in aggregator_col_names:
+            if stats_keys == aggregator_keys:
+                intersection.append(stats_keys)
+            elif (
+                stats_keys[0] == aggregator_keys[0]
+                and stats_keys[1] != aggregator_keys[1]
+            ):
+                type_differences.append([stats_keys, aggregator_keys])
+                aggregator_col_names.remove(aggregator_keys)
+                df_col_names.remove(stats_keys)
+
+    return intersection, type_differences, aggregator_col_names, df_col_names
```

Comment:
Use set logic for intersection and differences between two lists: https://realpython.com/python-sets/
much faster and well defined

Reply:
did use it in other places but here I did not because of the "elif" case and matching on only type and not name 

I could rewrite my structure so it works but is it necessary?